File: leetcode/p00142/p00142.go
package p00142

type ListNode struct {
	Val  int
	Next *ListNode
}

// Time Complexity: O(n)
// Space Complexity: O(1)
func detectCycle_pointers(head *ListNode) *ListNode {
	if head == nil {
		return nil
	}

	slow := head
	fast := head

	for fast.Next != nil && fast.Next.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next

		if slow == fast {
			slow2 := head
			for slow != slow2 {
				slow = slow.Next
				slow2 = slow2.Next
			}
			return slow
		}
	}

	return nil
}

// Time Complexity: O(n)
// Space Complexity: O(n)
func detectCycle_hashmap(head *ListNode) *ListNode {
	if head == nil {
		return nil
	}

	memo := map[*ListNode]int{}

	current := head
	for current.Next != nil {
		if _, exists := memo[current]; exists {
			return current
		}
		memo[current] = 1
		current = current.Next
	}

	return nil
}

File: leetcode/p00142/p00142.py
from typing import Optional

class ListNode:
    def __init__(self, val):
        self.val = val
        self.next = None

class Solution:
    # Time Complexity: O(n)
    # Space Complexity: O(n)
    def detectCycle_hashmap(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head == None:
            return None
        
        memo = {}
        current = head
        
        idx = 0
        while current.next != None:
            if current in memo:
                return current
            
            memo[current] = idx
            current = current.next
            idx += 1
    
        return None

    # Time Complexity: O(n)
    # Space Complexity: O(1)
    def detectCycle_pointers(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head == None:
            return None

        slow = head
        fast = head

        while fast.next != None and fast.next.next != None:
            slow = slow.next
            fast = fast.next.next
            
            if slow == fast:
                slow2 = head
                while slow != slow2:
                    slow = slow.next
                    slow2 = slow2.next
                return slow
        
        return None

if __name__ == '__main__':
    s = Solution()

    list = [ListNode(3), ListNode(2), ListNode(0), ListNode(-4)]
    for i in range(0, len(list)-1):
        list[i].next = list[i+1]
    list[-1].next = list[1]
    assert s.detectCycle_hashmap(list[0]) == list[1]
    assert s.detectCycle_pointers(list[0]) == list[1]

    list = [ListNode(3), ListNode(2), ListNode(0), ListNode(-4)]
    for i in range(0, len(list)-1):
        list[i].next = list[i+1]
    list[-1].next = list[0]
    assert s.detectCycle_hashmap(list[0]) == list[0]
    assert s.detectCycle_pointers(list[0]) == list[0]

    list = [ListNode(1)]
    assert s.detectCycle_hashmap(list[0]) == None
    assert s.detectCycle_pointers(list[0]) == None

    list = [ListNode(1), ListNode(1), ListNode(1), ListNode(1)]
    for i in range(0, len(list)-1):
        list[i].next = list[i+1]
    assert s.detectCycle_hashmap(list[0]) == None
    assert s.detectCycle_pointers(list[0]) == None